File: haushaltsbuch-domäne/src/main/java/com/github/haschi/haushaltsbuch/domäne/Haushaltsbuch.java
package com.github.haschi.haushaltsbuch.domäne;

import com.github.haschi.haushaltsbuch.api.Aggregatkennung;
import com.github.haschi.haushaltsbuch.api.ImmutableBeginneHaushaltsbuchführung;
import com.github.haschi.haushaltsbuch.api.ImmutableHauptbuchWurdeAngelegt;
import com.github.haschi.haushaltsbuch.api.ImmutableHaushaltsbuchführungBegonnen;
import com.github.haschi.haushaltsbuch.api.ImmutableJournalWurdeAngelegt;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.Buchung;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.ImmutableBuchung;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.ImmutableErstelleEröffnungsbilanz;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.ImmutableEröffnungsbilanzkontoErstellt;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.ImmutableKontenrahmenAngelegt;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.ImmutableKonto;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.Kontoart;
import com.github.haschi.haushaltsbuch.api.refaktorisiert.Spalte;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.eventsourcing.EventSourcingHandler;

import java.util.List;
import java.util.stream.Collectors;

import static org.axonframework.commandhandling.model.AggregateLifecycle.apply;

public class Haushaltsbuch
{
    @AggregateIdentifier
    Aggregatkennung id;

    public Haushaltsbuch()
    {

    }

    @CommandHandler
    public Haushaltsbuch(final ImmutableBeginneHaushaltsbuchführung anweisung)
    {
        apply(ImmutableHaushaltsbuchführungBegonnen.builder()
                        .id(anweisung.id())
                        .build());

        apply(ImmutableJournalWurdeAngelegt.builder()
                      .aktuelleHaushaltsbuchId(anweisung.id())
                      .build());

        apply(ImmutableKontenrahmenAngelegt.builder()
                      .addKonten(
                              ImmutableKonto.builder()
                                      .nummer("01")
                                      .bezeichnung("Sparbuch")
                                      .art(Kontoart.Aktiv)
                                      .build(),
                              ImmutableKonto.builder()
                                      .nummer("21")
                                      .bezeichnung("Geldbörse")
                                      .art(Kontoart.Aktiv)
                                      .build(),

                              ImmutableKonto.builder()
                                      .nummer("22")
                                      .bezeichnung("Bankkonto")
                                      .art(Kontoart.Aktiv)
                                      .build(),
                              ImmutableKonto.builder()
                                      .nummer("41")
                                      .bezeichnung("Bankkredit")
                                      .art(Kontoart.Passiv)
                                      .build(),
                              ImmutableKonto.builder()
                                      .nummer("51")
                                      .bezeichnung("Gehalt")
                                      .art(Kontoart.Ertrag)
                                      .build(),
                              ImmutableKonto.builder()
                                      .nummer("61")
                                      .bezeichnung("Miete")
                                      .art(Kontoart.Aufwand)
                                      .build(),
                              ImmutableKonto.builder()
                                      .nummer("80")
                                      .bezeichnung("Eröffnungsbilanzkonto")
                                      .art(Kontoart.Bilanz)
                                      .build())
                      .build());

        apply(ImmutableHauptbuchWurdeAngelegt.builder()
                      .haushaltsbuchId(anweisung.id())
                      .build());
    }

    @EventSourcingHandler
    public void falls(final ImmutableHaushaltsbuchführungBegonnen ereignis)
    {
        this.id = ereignis.id();
    }

    @CommandHandler
    public void erstelleEröffnungsbilanz(final ImmutableErstelleEröffnungsbilanz anweisung) {

        final List<ImmutableBuchung> buchungen =
                anweisung.inventar().vermögenswerte().stream()
                .map(vw -> ImmutableBuchung.builder()
                    .spalte(Spalte.haben)
                    .text(vw.beschreibung())
                    .betrag(vw.währungsbetrag())
                    .build())
                .collect(Collectors.toList());

        apply(ImmutableEröffnungsbilanzkontoErstellt.builder()
            .addAllBuchungen(buchungen)
              .build());
    }
}

File: core/src/main/kotlin/com/github/haschi/dominium/haushaltsbuch/core/model/Haushaltsbuch.kt
package com.github.haschi.dominium.haushaltsbuch.core.model

import com.github.haschi.dominium.haushaltsbuch.core.model.commands.BeginneHaushaltsbuchführung
import com.github.haschi.dominium.haushaltsbuch.core.model.events.EröffnungsbilanzkontoErstellt
import com.github.haschi.dominium.haushaltsbuch.core.model.events.HaushaltsbuchführungBegonnen
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Aggregatkennung
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Buchung
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Eröffnungsbilanzkonto
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Währungsbetrag
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.commandhandling.model.AggregateIdentifier
import org.axonframework.commandhandling.model.AggregateLifecycle
import org.axonframework.eventsourcing.EventSourcingHandler

class Haushaltsbuch
{

    @AggregateIdentifier
    private var id: Aggregatkennung? = null

    constructor()

    @CommandHandler
    constructor(anweisung: BeginneHaushaltsbuchführung)
    {
        AggregateLifecycle.apply(HaushaltsbuchführungBegonnen(anweisung.id))

        val inventar = anweisung.inventar

        val eigenkapital = Währungsbetrag(
                inventar.anlagevermoegen.summe.wert
                        .add(inventar.umlaufvermoegen.summe.wert)
                // .subtract(inventar.schulden().summe().wert())
        )

        val eröffnungsbilanz = Eröffnungsbilanzkonto(
                haben = listOf(
                        Buchung(
                                buchungstext = "Anlagevermögen (AV)",
                                betrag = inventar.anlagevermoegen.summe),
                        Buchung(
                                buchungstext = "Umlaufvermögen (UV)",
                                betrag = inventar.umlaufvermoegen.summe)),
                soll = listOf(
                        Buchung(
                                buchungstext = "Eigenkapital (EK)",
                                betrag = eigenkapital)))

        //                .addSoll(Buchung.builder()
        //                                 .buchungstext("Fremdkapital (FK)")
        //                                 .betrag(inventar.schulden().summe())
        //                                 .build())

        AggregateLifecycle.apply(EröffnungsbilanzkontoErstellt(eröffnungsbilanz))
    }

    @EventSourcingHandler
    fun falls(ereignis: HaushaltsbuchführungBegonnen)
    {
        id = ereignis.id
    }
}
